### PR TITLE
added verify to Report, replacing verify_report from libenclave

### PR DESCRIPTION
**Pull request by Goirad**
Added an inherent method that allows for local verification of the report

### CODE REVIEW

File: sgx-isa/src/lib.rs [hunk 1]
```diff
@@ -576,6 +576,21 @@ impl Report {
         // unwrap ok, `out` is the correct number of bytes
         Report::try_copy_from(&out.0).unwrap()
     }
+
+    #[cfg(all(feature = "sgxstd", target_env = "sgx"))]
+    pub fn verify<F, R>(&self, check_mac: F) -> R
+        where F: FnOnce(&[u8], &[u8], &mut [u8]) -> R {
+	let req = Keyrequest {
```
Comment (jethrogb) [inline sgx-isa/src/lib.rs]: Indentation issue. Try `rustfmt`.

File: sgx-isa/src/lib.rs [hunk 2]
```diff
@@ -576,6 +576,21 @@ impl Report {
...
+	    keyname: Keyname::Report as u16,
+	    keyid: self.keyid,
+	    ..Default::default()
+	};
+	let key = req.egetkey().expect("Couldn't get report key");
+        let mut out = [0 as u8; 16];
+        check_mac(&key,
+                  unsafe{::std::slice::from_raw_parts(self as *const _ as *const u8, 384)},
+                  &mut out)
```
Comment (jethrogb) [inline sgx-isa/src/lib.rs]: This should be self.mac

File: sgx-isa/src/lib.rs [hunk 3]
```diff
@@ -576,6 +576,21 @@ impl Report {
         // unwrap ok, `out` is the correct number of bytes
         Report::try_copy_from(&out.0).unwrap()
     }
+
+    #[cfg(all(feature = "sgxstd", target_env = "sgx"))]
+    pub fn verify<F, R>(&self, check_mac: F) -> R
+        where F: FnOnce(&[u8], &[u8], &mut [u8]) -> R {
```
Comment (jethrogb) [inline sgx-isa/src/lib.rs]: Please use the function signature I suggested (with fixed-size arrays)

File: sgx-isa/src/lib.rs [hunk 4]
```diff
@@ -576,6 +580,24 @@ impl Report {
         // unwrap ok, `out` is the correct number of bytes
         Report::try_copy_from(&out.0).unwrap()
     }
+
+    #[cfg(all(feature = "sgxstd", target_env = "sgx"))]
+    pub fn verify<F, R>(&self, check_mac: F) -> R
```
Comment (VardhanThigle) [inline sgx-isa/src/lib.rs]: small suggestion:
Would it be possible to add documentation here to tell users that check_mac should prevent a timing attack?
Another possibility could be that we accept a mac call back and we equate macs of mac.

Comment (jethrogb) [reply sgx-isa/src/lib.rs]: Let's go with docs

Comment (Goirad) [reply sgx-isa/src/lib.rs]: Just `// check_mac should prevent timing attacks` ?

Comment (jethrogb) [reply sgx-isa/src/lib.rs]: Well if you're documenting, might as well write the full docs for the function. Also, you should call out the comparison specifically as an area of concern.

PS. Doc comments are triple-slash

Comment (Goirad) [reply sgx-isa/src/lib.rs]: Added comments

**Pull request merged**